src/Analyzer/QueryPlanAnalyzerMysql.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+declare(strict_types=1);
+
+namespace staabm\PHPStanDba\Analyzer;
+
+use mysqli;
+use PDO;
+use PHPStan\ShouldNotHappenException;
+use staabm\PHPStanDba\QueryReflection\QueryReflection;
+
+final class QueryPlanAnalyzerMysql
+{
+    /**
+     * number of unindexed reads allowed before a query is considered inefficient.
+     */
+    public const DEFAULT_UNINDEXED_READS_THRESHOLD = 100000;
+
+    /**
+     * @var PDO|mysqli
+     */
+    private $connection;
+
+    /**
+     * @param PDO|mysqli $connection
+     */
+    public function __construct($connection)
+    {
+        $this->connection = $connection;
+    }
+
+    /**
+     * @param non-empty-string $query
+     */
+    public function analyze(string $query): QueryPlanResult
+    {
+        if ($this->connection instanceof PDO) {
+            $stmt = $this->connection->query('EXPLAIN '.$query);
+
+            // @phpstan-ignore-next-line
+            return $this->buildResult($stmt);
+        } else {
+            $result = $this->connection->query('EXPLAIN '.$query);
+            if ($result instanceof \mysqli_result) {
+                return $this->buildResult($result);
+            }
+        }
+
+        throw new ShouldNotHappenException();
+    }
+
+    /**
+     * @param \IteratorAggregate<array-key, array{key: string|null, rows: positive-int, table: ?string}> $it
+     */
+    private function buildResult($it): QueryPlanResult
+    {
+        $result = new QueryPlanResult();
+        $allowedUnindexedReads = QueryReflection::getRuntimeConfiguration()->getNumberOfAllowedUnindexedReads();
+
+        if (false === $allowedUnindexedReads) {
+            throw new ShouldNotHappenException();
+        }
+
+        foreach ($it as $row) {
+            // we cannot analyse tables without rows -> mysql will just return 'no matching row in const table'
+            if (null === $row['table']) {
+                continue;
+            }
+
+            if (null === $row['key']) {
+                $result->addRow($row['table'], QueryPlanResult::NO_INDEX);
+            } else {
+                if (true === $allowedUnindexedReads && $row['rows'] > self::DEFAULT_UNINDEXED_READS_THRESHOLD) {
+                    $result->addRow($row['table'], QueryPlanResult::NOT_EFFICIENT);
+                }
+                if (\is_int($allowedUnindexedReads) && $row['rows'] > $allowedUnindexedReads) {
+                    $result->addRow($row['table'], QueryPlanResult::NOT_EFFICIENT);
+                }
+            }
+        }
+
+        return $result;
+    }
```

Comment:
@tpetry may I ask you about feedback regarding how to interpret a mysql explain result properly?

I am working on adding perf indication within phpstandba with this PR.
my goal is to emit phpstan errors when queries seem inefficient, indicated by the mysql explain result of said queries.

you can see this feature "in action" in the 'Files Changed'-Tab of https://github.com/composer/packagist/pull/1304/files

example:

![grafik](https://user-images.githubusercontent.com/120441/170224578-14a8a9d5-7879-498c-99e2-a928f3fba498.png)

Reply:
thx for the feedback. 

my point is that it is helpful when the tool detects queries which don't use a index. the developer might ignore/baseline this error after analyzing the "false-positive". the feature is more about making the developer aware of queries not using indizes - and confirm these.

Reply:
I added a threshold, so we only warn about non-indexed and full-table-scan problems for tables with more then 1000 rows